File: Applied Course/4.Problem Solving/14.Maths & Bit Manipulation/4.Maximum Product of word Lengths.py
# https://leetcode.com/problems/maximum-product-of-word-lengths
# Logic-1
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        def no_common_letters(word1,word2):
            for c in word1:
                if c in word2:
                    return False
            return True
        n=len(words)
        max_val=0
        for i in range(n):
            for j in range(i+1,n):
                if no_common_letters(words[i],words[j]):
                    max_val=max(max_val,len(words[i])*len(words[j]))
        return max_val
# Logic-2
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        n = len(words)
        masks = [0] * n
        lens = [0] * n
        bit_number = lambda ch : ord(ch) - ord('a')
        
        for i in range(n):
            bitmask = 0
            for ch in words[i]:
                # add bit number bit_number in bitmask
                bitmask |= 1 << bit_number(ch)
            masks[i] = bitmask
            lens[i] = len(words[i])
            
        max_val = 0
        for i in range(n):
            for j in range(i + 1, n):
                if masks[i] & masks[j] == 0:
                    max_val = max(max_val, lens[i] * lens[j])
        return max_val

File: Applied Course/4.Problem Solving/14.Maths & Bit Manipulation/4.Maximum Product of word Lengths.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool no_common_letters(string word1,string word2)
    {
        for(char c:word1)
            if(word2.find(c)!=string::npos)
                return false;
        return true;
    }
    int maxProduct(vector<string>& words) {
        int n=words.size();
        int max_val=0;
        for(int i=0;i<n;i++)
        {
            for(int j=i+1;j<n;j++)
            {
                if(no_common_letters(words[i],words[j]))
                {
                    int len=words[i].length()*words[j].length();
                    max_val=max(max_val,len);   
                }
            }
        }
        return max_val;
    }
};
int main()
{
    // https://leetcode.com/problems/maximum-product-of-word-lengths/
    return 0;
}